tests/apps/test_train.py

```diff
@@ -0,0 +1,213 @@
+    are not the same, (ii) they are rescaled by the correct factor."""
+    adj = np.ones((10, 10))
+    n_mean = 5
+
+    r1 = train.parametrization.rescale_adjacency(adj, n_mean, True)
+    r2 = train.parametrization.rescale_adjacency(adj, n_mean, False)
+    s1 = rescale_tor(adj, n_mean)
+    s2 = rescale(adj, n_mean)
+
+    assert not np.allclose(r1, r2)
+    assert np.allclose(r1, s1 * adj)
+    assert np.allclose(r2, s2 * adj)
+
+
+def test_A_to_cov():
+    """Test if A_to_cov returns the correct covariance matrix for a simple fixed example"""
+    x = np.sqrt(0.5)
+    A = np.array([[0, x], [x, 0]])
+    cov = train.parametrization.A_to_cov(A)
+    target = np.array(
+        [
+            [1.5, 0.0, 0.0, 1.41421356],
+            [0.0, 1.5, 1.41421356, 0.0],
+            [0.0, 1.41421356, 1.5, 0.0],
+            [1.41421356, 0.0, 0.0, 1.5],
+        ]
+    )
+    assert np.allclose(cov, target)
+
+
+@pytest.mark.usefixtures("adj", "params")
+@pytest.mark.parametrize("dim", [7, 9])  # used in the adj fixture to determine number of modes
+@pytest.mark.parametrize("n_mean", [4, 6])
+class TestVGBS:
+    """Tests for the class ``train.VGBS``"""
+
+    def test_W(self, adj, n_mean, params):
+        """Test that the W method correctly gives the diagonal matrix of embedded parameters"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        W = gbs.W(params)
+        assert np.allclose(np.diag(W), embedding(params))  # check that diagonal equals embedding
+        assert np.allclose(np.diag(np.diag(W)), W)  # check that matrix is diagonal
+
+    def test_A(self, adj, n_mean, params):
+        """Test that the A and _A_scaled methods correctly return the WAW matrix of the A_init
+        and A_init_scaled adjacency matrices, respectively, and that these outputs are different
+        (due to the scaling)"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        A = gbs.A(params)
+        A_scaled = gbs._A_scaled(params)
+        W = gbs.W(params)
+
+        assert np.allclose(A, W @ adj @ W)
+        assert np.allclose(A_scaled, W @ gbs.A_init_scaled @ W)
+        assert not np.allclose(A, A_scaled)
+
+    def test_generate_samples(self, adj, n_mean, monkeypatch):
+        """Test that _generate_samples correctly dispatches between torontonian and hafnian
+        sampling based upon whether threshold=True or threshold=False"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+
+        with monkeypatch.context() as m:
+            m.setattr(thewalrus.samples, "torontonian_sample_state", lambda *args, **kwargs: 0)
+            s_threshold = gbs._generate_samples(gbs.A_init_scaled, 10)
+
+        gbs.threshold = False
+
+        with monkeypatch.context() as m:
+            m.setattr(thewalrus.samples, "hafnian_sample_state", lambda *args, **kwargs: 1)
+            s_pnr = gbs._generate_samples(gbs.A_init_scaled, 10)
+
+        assert s_threshold == 0
```

Comment:
I'm confused by this test. Is the idea that we're testing whether we are making the right call to the walrus function? I thought the output of `_generate_samples` was a sample, so how come we're testing if it's `0` or `1`? Not saying there's a problem with the test, only that it's hard to follow and therefore to review

Reply:
(Minor suggestion): How about creating two small test cases by splitting this one up for further separation? `test_generate_samples_torontonian`, `test_generate_samples_hafnian`